### PR TITLE
[SMTChecker] Add callstack to model report

### PR DESCRIPTION
**Pull request by leonardoalt**
Split #6244 

For the following code, part of the report will be
```
pragma experimental SMTChecker;

contract C
{
	function f(uint x) public pure returns (uint) {
		assert(x > 0);
		return x + 1;
	}
	function g(uint x) public pure returns (uint) {
		return f(x);
	}
	function h(uint x) public pure {
		uint y = g(x);
	}
}
```

```
a.sol:6:3: Warning: Assertion violation happens here
		assert(x > 0);
		^-----------^
  for:
   = 0
  x = 0
  y = 0

Callstack: 
a.sol:13:12: 
		uint y = g(x);
		         ^--^
a.sol:10:10: 
		return f(x);
		       ^--^
```

Did not add tests because this is a secondary location.

### CODE REVIEW

**Comment by leonardoalt** [comment]
Just noticed this needs to be reversed :p

**Comment by leonardoalt** [comment]
Reversed. Now the output is:
```
a.sol:6:3: Warning: Assertion violation happens here
		assert(x > 0);
		^-----------^
  for:
   = 0
  x = 0
  y = 0

Callstack: 
a.sol:10:10: 
		return f(x);
		       ^--^
a.sol:13:12: 
		uint y = g(x);
		         ^--^
```

File: libsolidity/formal/SMTChecker.cpp [hunk 1]
```diff
@@ -1165,18 +1169,21 @@ void SMTChecker::checkCondition(
...
 	if (m_arrayAssignmentHappened)
-		loopComment +=
+		extraComment +=
 			"\nNote that array aliasing is not supported,"
 			" therefore all mapping information is erased after"
 			" a mapping local variable/parameter is assigned.\n"
 			"You can re-introduce information using require().";
 
+	SecondarySourceLocation secondaryLocation{};
+	secondaryLocation.append(extraComment, SourceLocation());
```
Comment (erak) [inline libsolidity/formal/SMTChecker.cpp]:
```suggestion
	secondaryLocation.append(extraComment, SourceLocation{});
```

File: libsolidity/formal/SMTChecker.cpp [hunk 2]
```diff
@@ -1195,19 +1202,28 @@ void SMTChecker::checkCondition(
 
 			for (auto const& eval: sortedModel)
 				modelMessage << "  " << eval.first << " = " << eval.second << "\n";
-			m_errorReporter.warning(_location, message.str(), SecondarySourceLocation().append(modelMessage.str(), SourceLocation()).append(loopComment, SourceLocation()));
+			m_errorReporter.warning(
+				_location,
+				message.str(),
+				SecondarySourceLocation().append(
+					modelMessage.str(),
+					SourceLocation()
```
Comment (erak) [inline libsolidity/formal/SMTChecker.cpp]:
```suggestion
					SourceLocation{}
```

Comment (erak) [reply libsolidity/formal/SMTChecker.cpp]: The same could be applied in line 1208.

**Review by erak — DISMISSED** [review]
Looks good! (apart from the tiny suggestions that do not necessarily need to be applied)

**Comment by leonardoalt** [comment]
@erak updated

**Review by erak — DISMISSED** [review]

File: liblangutil/Exceptions.h [hunk 3]
```diff
@@ -28,17 +28,20 @@
 #include <memory>
 #include <libdevcore/Exceptions.h>
 #include <libdevcore/Assertions.h>
+#include <libdevcore/CommonData.h>
 #include <liblangutil/SourceLocation.h>
 
+using namespace dev;
```
Comment (chriseth) [inline liblangutil/Exceptions.h]: No `using namespace` inside header.

File: liblangutil/Exceptions.h [hunk 4]
```diff
@@ -108,6 +111,11 @@ class SecondarySourceLocation
 		infos.emplace_back(_errMsg, _sourceLocation);
 		return *this;
 	}
+	SecondarySourceLocation& append(SecondarySourceLocation const& _other)
+	{
+		infos += _other.infos;
```
Comment (chriseth) [inline liblangutil/Exceptions.h]: move?

**Comment by leonardoalt** [comment]
@erak @chriseth updated & rebased

**Review by erak — APPROVED** [review]

**Review by chriseth — APPROVED** [review]

**Pull request merged**